Aggregate health status in unified resource cache

GavinFrazar opened the pull request:
Part of:
- #20544

This should also fix:
- https://github.com/gravitational/teleport/issues/46937

Web UI and Teleterm will be updated in separate PRs to leverage the URC change and fetch all health statuses.

The health status for a resource with multiple servers, as in a database proxied by multiple agents, will now be aggregated in the unified resource cache.
If there are multiple statuses and the statuses are not all the same, then the aggregate status will be set to "mixed".
It is then the client's responsibility to list the servers for a "mixed" status resource if they want details about individual resource server health.

The purpose is to allow the web ui to efficiently fetch all health statuses for all resources in the unified resources view.

GavinFrazar committed:
Aggregate health status in unified resource cache

The health status for a resource with multiple servers, as in a database
proxied by multiple agents, will now be aggregated in the unified
resource cache. If there are multiple statuses and the statuses are not
all the same, then the aggregate status will be set to "mixed".
It is then the client's responsibility to list the servers for a "mixed"
status resource if they want details about individual resource server
health.

greedy52 commented on api/proto/teleport/legacy/client/proto/authservice.proto:
```diff
@@ -2053,12 +2053,26 @@ message PaginatedResource {
 
+  // TargetHealthStatuses is a list of health statuses for the resource. There
+  // can be multiple statuses if multiple agents proxy the same resource.
+  repeated TargetHealthStatus TargetHealthStatuses = 17 [(gogoproto.jsontag) = "target_health_statuses,omitempty"];
```
alternatively maybe use a message similar to `types.TargetHealthGroups` but only counts? like `enriched_resource.target_counts_by_health_status.healthy = 5` (i am not at naming, just an example)

greedy52 replied:
the count struct/message is more deterministic than a list that can be any size like `[0,0,1,1,0,0,1,1,2,1,2,1,1,1,1,0]` and the list requires enum conversion. (i know this example is unrealistic but you never know =p)

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -128,31 +129,27 @@ func (cfg *UnifiedResourceCacheConfig) CheckAndSetDefaults() error {
+		old.sortKey = sortKey
 	}
-	c.resources[key] = resource
+	if c.resources[key] == nil {
+		c.resources[key] = newResourceCollection(resource.GetKind(), sortKey)
+	}
+	c.resources[key].put(resource)
```
nit: maybe we can reduce number of `c.resources[key]` here, currently it does it 3-4 times per `putLocked`.

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -173,9 +170,12 @@ func (c *UnifiedResourceCache) deleteLocked(res types.Resource) error {
-	sortKey := makeResourceSortKey(resource)
-	c.deleteSortKey(sortKey)
-	delete(c.resources, key)
+	if resource.isSingular() {
+		c.deleteSortKey(resource.sortKey)
+		delete(c.resources, key)
+	} else {
+		resource.removeLocked(res)
+	}
```
i don't know if it's possible but should we handle the case where `res` being removed is not part of cached list of `servers`?

greedy52 replied:
for server type, say database server, `isSingular` checks:
```
func (c *resourceCollection) isSingular() bool {
	return len(c.servers) <= 1
}
```
but is it possible if `c.servers[0]` is not the one getting removed here? say OpDelete happens twice for the 2nd last server. if it's unlikely then don't worry about it.

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -244,13 +244,13 @@ func (c *UnifiedResourceCache) iterateItems(ctx context.Context, start string, s
-					if len(kinds) == 0 || c.itemKindMatches(r, kindsMap) {
-						items = append(items, iteratedItem{key: item.Key, resource: r})
+					if len(kinds) == 0 || c.itemKindMatches(collection.getKind(), kindsMap) {
+						items = append(items, iteratedItem{key: item.Key, collection: collection.copyLocked()})
```
it's quite expensive to copy here as the resource will be copied again outside after applying filter. any way to avoid it?

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -998,6 +969,160 @@ type resource interface {
+		return c.latest, match, trace.Wrap(err)
+	}
+
+	for _, r := range c.servers {
```
we should think this through, and define clearly what's the behavior when using predicate on health status for an unified resource.

for example, if one agent is healthy but the other is not, and you search for `health.status == "healthy"`, you would see a tile with warning shows up.

Personally, i am still bias towards a calculated single status that aggregates from all heartbeats. Let me know what you think.

greedy52 replied:
> I think that makes sense - you search for healthy and there is a healthy server for that db, we just group them all together in one tile. You see a warning because not all servers for that db are healthy, which also seems to make sense

Do we show the healthy server on UI? the side panel only show problematic host IDs. it would be confusing to the user why it showed up.

greedy52 replied:
the match function is still against individual servers or should it against the aggregated status? if later we can also simplify the copying logic

greedy52 reviewed (commented):
first pass

greedy52 commented on lib/healthcheck/worker.go:
```diff
@@ -361,15 +361,23 @@ func (w *worker) getThreshold(cfg *healthCheckConfig) uint32 {
+	checkWord := pluralize(w.lastResultCount, "check")
 	if w.lastResultErr == nil {
-		msg := fmt.Sprintf("%d health checks passed", w.lastResultCount)
+		msg := fmt.Sprintf("%d health %v passed", w.lastResultCount, checkWord)
```
nit: i'm good with our own helper. I see that we already importing this lib so just curious if some of these helpers are useful
https://pkg.go.dev/github.com/dustin/go-humanize@v1.0.1/english

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -998,6 +968,138 @@ type resource interface {
+}
+
+func (d *aggregatedDatabase) GetTargetHealth() types.TargetHealth {
+	out := d.DatabaseServer.GetTargetHealth()
+	out.Status = string(d.status)
```
Since using aggregated status, do we still need other information from `d.DatabaseServer.GetTargetHealth()`?

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -998,6 +968,138 @@ type resource interface {
+
+func (c *databaseServerCollection) remove(r types.Resource) bool {
+	delete(c.servers, r.GetMetadata().Description)
+	c.aggregate = &aggregatedDatabase{
+		DatabaseServer: c.aggregate.DatabaseServer,
+		status:         aggregateHealthStatuses(c.servers),
+	}
+	return len(c.servers) == 0
```
should we double-check if `c.aggregate.DatabaseServer` is the one being removed? or just get any server that's still in `c.servers`

greedy52 commented on lib/web/servers.go:
```diff
@@ -143,17 +143,32 @@ func (h *Handler) clusterDatabaseGet(w http.ResponseWriter, r *http.Request, p h
 	}
 
-	dbServer, err := fetchDatabaseServerByDatabaseName(r.Context(), clt, r, databaseName)
+	dbServers, err := fetchDatabaseServersWithName(r.Context(), clt, r, databaseName)
```
nit: any benefit using `fetchDatabaseServersWithName` versus using the unified resource api here? maybe we should switch to the unified resource API on Web UI side.

GavinFrazar committed:
Update BenchmarkListUnifiedResources

GavinFrazar committed:
Improve aggregate health status cache performance

* reduce support for health predicate expressions to only support
  expressions on the "health.status" field. This reduces copying of the
  entire TargetHealth struct and eliminates the dynamic json tag lookup
  for health status matching.
* add GetTargetHealthStatus and SetTargetHealthStatus. This also reduces
  copying of the entire TargetHealth struct. It's also more convenient,
  as most of the time we just care about getting/setting health status
  field.

greedy52 commented on lib/services/unified_resource.go:
```diff
@@ -940,6 +908,150 @@ type resource interface {
+// aggregatedDatabase wraps a database server with aggregated health status.
+// This type exists to avoid cloning the resource unnecessarily, yet still
+// prevent data races.
+type aggregatedDatabase struct {
```
gavin is on PTO this week so i am trying to pick this up. how about something like this using generic?
[e3e56bb](https://github.com/gravitational/teleport/pull/54868/commits/e3e56bb2b7698a29c6c39eb52c3870889ba2f262)

I have to drop `Copy` though as it does play very well.

rosstimothy reviewed (approved):
Thank you for your patience and effort on this @greedy52 and @GavinFrazar!
